text/3604-derive_c-enum_integer_conversions.md

```diff
@@ -0,0 +1,185 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+
+I want to propose a **TryFrom<{Integer}>** and **Into<{Integer}>** derive macro becomes available for all C-style enums. C-style enums are a subset of all integers, which is why C-style Enums can be cast to integers using the ``as {usize/u8/isize, etc..}`` syntax. I propose that we create a derive macro, that automatically allows C-Style Enums to be cast to Integers using ``.into()``, not just ``as``. Since only _some_ Integers can be turned into a C-Style Enum, I propose that a TryFrom<{Integer}> derive macro can be made for them, to allow easy and more generic conversion between a C-Style Enum and Integers.
+
+## \#[derive(TryFromInt)] implementation
+The logic of the derive macro would look something like this
+```
+make a TryFrom impl block for all of the integer types (u8,u16, etc..)
+make an impl similar to this for all unsigned ints
+impl TryFrom<u{8,16,..}> for InputtedEnum {
+  type Error = (); //only one point of failure for conversion, so use unit
+  fn try_from(value: u{8,16,...}) -> Result<InputtedEnum, ()>
+    //this makes it so you can`t match a value for a field that has a value greater that u{8,16,...}::MAX
+    //e.g. if an enum had a field with a value of 256, but we were implementing TryFrom for a u8, we would
+    //have to do bounds checking and change the function body to only match on fields with a value less than
+    //u{8,16,..}::MAX and more than u{8,16,...}::MIN
+    let value = value as usize;
+    match value { 
+        InputtedEnum::Field1 as usize => Ok(InputtedEnum::Field1),
+        InputtedEnum::FieldN as usize => Ok(InputtedEnum::FieldN),
+        none_of_those_fields => Err(()) //no fields have inputted value, conversion failed
+    }
+  }
+}
+```
+I\`m sure someone will be able to think of some cool optimizations for this, but this is plenty fast already
+
+## \#[derive(IntoInt)] implmentaion
+This one is really simple, its just using the `as` conversion inside the function, to convert the inputted enum to any number. But this allows it to be used inside generics where Into<{Integer}> is a trait bound, among others.
+```
+impl Into<{Integer}> for InputtedEnum {
+  fn into(self) -> {Integer} {
+    self as {Integer}
+  }
+}
+```
```

Comment:
For field-less (but not unit-only) enums, this will produce incorrect results. Consider this:
```
enum Example {
  Foo
  Bar {}
  Baz()
}
```
`Baz` as `{integer}` will produce not the discriminant of `Baz`, but rather the address of its implicit constructor.